### DISCUSSION TITLE
ddp: how to combine multi-gpus outputs like "training_step_end" which is only used in dp/ddp2?

### DISCUSSION THREAD

**Discussion by jiyt17** [Lightning-AI/pytorch-lightning#11265] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
My question is like title. Thank you!

**Comment by rohitgr7** [comment, marked as answer]
checkout the example here: https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#training-with-dataparallel

**Reply by jiyt17** [reply]
But what about combine multi-gpus outputs in **ddp**?

**Reply by rohitgr7** [reply]
you can use all_gather for that:
https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#all-gather

here is one example I posted for prediction, you can apply that for any case.
https://github.com/PyTorchLightning/pytorch-lightning/discussions/9259#discussioncomment-1858041

**Reply by jiyt17** [reply]
Thank you, I'll have a try

**Reply by WouterDurnez** [reply]
I actually posted the same question before, and got it to work thanks to the responses. Using `all_gather`worked, provided that all my metrics were tensors. In my case, that meant I had to convert some scalars to the tensor format, after which `all_gather` yielded concatenated metrics.
